java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestGenerationUtils.java

```diff
@@ -152,16 +163,280 @@ public static WorkspaceEdit generateTests(List<Object> arguments, IProgressMonit
+        final IJavaProject javaProject = unit.getJavaProject();
+        final List<TestKind> testFrameworksInProject = TestKindProvider.getTestKindsFromCache(javaProject);
+        final TestKind testKind = determineTestFramework(new HashSet<>(testFrameworksInProject));
+        if (testKind == null) {
+            return null;
+        }
+
+        final IClasspathEntry testEntry = getTestClasspathEntry(unit);
+        final String testFullyQualifiedName = getTestFullyQualifiedName(typeBinding, javaProject, testEntry);
+        if (testFullyQualifiedName == null) {
+            return null;
+        }
+        final List<String> methodsToTest = getMethodsToTest(typeBinding);
+        if (methodsToTest == null || methodsToTest.size() == 0) {
+            return null;
+        }
+
+        final ICompilationUnit testUnit = getTestCompilationUnit(javaProject, testEntry, testFullyQualifiedName);
+        return createTextEditFromSourceFile(testKind, testUnit, methodsToTest, cursorOffset);
+    }
+
+    private static WorkspaceEdit createTextEditFromSourceFile(TestKind kind, ICompilationUnit testUnit,
+            List<String> methodsToTest, int cursorPosition) throws CoreException {
+        if (testUnit.exists()) {
+            final IType[] types = testUnit.getAllTypes();
+            if (types.length == 0) {
+                // an empty java file
+                return getTestCompilationUnit(kind, testUnit, methodsToTest);
+            }
+
+            final CompilationUnit root = (CompilationUnit) TestSearchUtils.parseToAst(testUnit,
+                false /* fromCache */, new NullProgressMonitor());
+
+            if (root == null) {
+                return null;
+            }
+
+            IType type = testUnit.findPrimaryType();
+            if (type == null) {
+                type = types[0];
+            }
+
+            final ASTNode typeNode = root.findDeclaringNode(type.getKey());
+            if (!(typeNode instanceof TypeDeclaration)) {
+                return null;
+            }
+
+            final ITypeBinding binding = ((TypeDeclaration) typeNode).resolveBinding();
+            if (binding == null) {
+                return null;
+            }
+
+            IJavaElement insertPosition = null; // Insert to the last by default.
+            try {
+                insertPosition = CodeGenerationUtils.findInsertElement(type, cursorPosition);
+            } catch (Throwable e) {
+                // ignore if the upstream does not support insert position preference.
+            }
+
+            return createEditToExistingTestFile(root, kind, methodsToTest, (TypeDeclaration) typeNode,
+                    binding, insertPosition);
+        } else {
+            return createAndGetNewTestCompilationUnit(kind, testUnit, methodsToTest);
+        }
+    }
+
+    /**
+     * Get a compilation unit change without creating.
+     */
+    private static WorkspaceEdit getTestCompilationUnit(TestKind kind, ICompilationUnit testUnit,
```

Comment:
the method name is confusing.

getTestCompilationUnit(...)  -> addTestClassToExistingFile(...)